capplets/common/mate-theme-apply.c

```diff
@@ -48,13 +49,22 @@ void
 mate_meta_theme_set (MateThemeMetaInfo *meta_theme_info)
 {
   GSettings *interface_settings;
+  GSettings *interface_gnome_settings;
   GSettings *marco_settings;
   GSettings *mouse_settings;
   GSettings *notification_settings = NULL;
   gchar *old_key;
   gint old_key_int;
+  int ret;
+
+  /*Find out if we are running under wayland and have xwayland
+   *Sending signal 0 with killall doesn't kill the process but
+   *the return value tells us whether the process exists
+   */
+  ret = system ("killall -0 -e Xwayland");
```

Comment:
I just tested that change, and unfortunately I could not get it to work. The reason is that with system() we get the return value of the program run, and that's what we are using to distinguish the wayland session from an x11 session while running under Xwayland.

With `g_spawn_command_line_async() `we don't. Since finding nothing on running `killall -0 -e <anything>` does not seem to be considered an error, using GError doesn't seem to give a way of distinguishing this either. While this function is a gboolean returning TRUE or FALSE, it seems to return TRUE unconditionally when running 
`g_spawn_command_line_async("killall -0 -e Xwayland", NULL);`
I tested that by comparing behavior in both wayland and x11 sessions

A different issue cropped up trying to use` g_spawn_command_line_async()` to start mate-appearance-properties from caja in https://github.com/mate-desktop/caja/pull/1771
 if using mate-control-center 1.28.0 or current git master, we cannot open mate-appearance-properties without passing `GDK_BACKEND=x11` as an environmental variable. We can get by without it entirely in caja if this PR is applied to mate-control-center, but that would cause caja to have different behavior in wayland depending on which version of m-c-c is installed, shortly after a major release. Not a hard dependency due to the fallback, but something that might generate bug reports.

 `g_spawn_command_line_async()` does not seem to be able to pass environmental variables, and neither do the similar commands such as those that underlie `caja_launch_application_from_command`. That one can be made to work by setting it to open a command in terminal, but then we have that showing on the desktop plus all the code to run a totally unneeded terminal. Three things worked there:  `g_spawn_command_line_async()` or `caja_launch_application_from_command()` with this PR applied to m-c-c and not trying to pass any environmental variables, or `caja_launch_application_from_command()` setting terminal usabe to TRUE, which allowed  `GDK_BACKEND=x11` to be passed and the background page of this capplet to open.

I agree that it would be good to avoid `system()` as it is heavy, but this seems to be hard to avoid here. In this case, an alternative mode of detecting that we are in a wayland session from an app running in Xwayland might be able get us around using `system()` which is rather heavy for this job.  In the caja case by comparison, the fork etc should be less of a concern as mate-appearance-properties being launched from there is a whole new toplevel program anyway. There too though we have to run two checks, there too I tested g_spawn_command_line_async() and the checks didn't work.

Thoughts?

Reply:
Since https://github.com/mate-desktop/mate-control-center/pull/749/commits/3835297e5add6c2d3327d5813b861939e41d1713 the appearance capplet can now run under native wayland. That means we no longer need that system call to detect the session so
https://github.com/mate-desktop/mate-control-center/pull/749/commits/4832386262c8fbadf6cce5bacb0b571b15b99338

removes it, and this should make lighter code